drivers: spi: stm32h7 Fix SPI DMA 16 bits frames

dgastonochoa opened the pull request:
This is a proposal to add tests for 16 bits frames with DMA. These tests have revealed 2 problems:

* 16 bits frames DMA seems to not work (the 16 bits frames tests don't pass). The driver doesn't update buffer pointers correctly, it only updates them as if the frame size was always 8 bits.
* The driver uses an internal buffer to send zeroes when tx is finished or NULL, and to dump data from rx when the rx buffer is finished or NULL. The problem is that this buffer is not in a nocache area, os its value is not necessarily zero for DMA. This is visible only in the 16 bits tests (not the 8 bits ones), and makes DMA to fail. 

This fixes the problem mentioned in #61904

This PR also simplifies the test source code (`spi.c`) by removing specific SPI configurations in it in favor of configuration files such as `*.conf`, `*.overlay` and `testcase.yaml`. 

Test plan
=========

Connect an STM32H7 board to the PC with pins D11 and D12 connected.

Open a minicom or equivalent terminal to read the test report.

Execute the following commands:

```
rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h753zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_dma.loopback
west flash

rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h753zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_16bits_frames_dma.loopback
west flash
```

And verify all tests pass.

dgastonochoa commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -88,7 +94,12 @@ static const struct mem_region nocache_mem_regions[] = {
  */
-uint32_t dummy_rx_tx_buffer;
+#ifdef CONFIG_NOCACHE_MEMORY
+static __aligned(32) uint32_t dummy_rx_tx_buffer
```
Running the 16 bit frames tests has revealed a bug in the SPI driver: the dummy_rx_tx_buffer is not in a nocache region. When DMA is enabled, this means that his variable's value will not necessarily be zero. This is a problem as the dummy buffer should always be 0, so SPI sends zeroes whenever required (there are a number of situations in which this is required, see spi.h interface. In summary, when rx is finished or NULL but tx is not, and vice-versa).

The solution to this problem is setting the variable in a nocache region, but this can only be done when CONFIG_NOCACHE_MEMORY. When it is not, the nocache regions are defined in devicetree, but are unknown to this driver. Furthermore, there could be several nocache regions declared there, and it's not clear which one should be used, even if it could be accessed.

To solve this, if CONFIG_NOCACHE_MEMORY is not enabled, don't place the variable in a nocache area. Instead, keep cache coherency manually.

erwango replied:
A summary of this comment would be valuable in code.

dgastonochoa replied:
I added some comments. Hopefully they are helpful.

dgastonochoa commented on drivers/spi/Kconfig.stm32:
```diff
@@ -21,6 +21,7 @@ config SPI_STM32_INTERRUPT
 config SPI_STM32_DMA
 	bool "STM32 MCU SPI DMA Support"
 	select DMA
+	select CACHE_MANAGEMENT
```
This is required to use the functions used to keep cache coherency. See explanation in the dummy_rx_tx_buffer variable.

dgastonochoa commented on tests/drivers/spi/spi_loopback/Kconfig:
```diff
@@ -6,3 +6,20 @@ source "Kconfig.zephyr"
+config DT_DEFINED_NOCACHE
+	bool "Enable this if nocache regions are defined in devicetree"
+
+if DT_DEFINED_NOCACHE
```
Nocache buffers can be obtained with CONFIG_NOCACHE_MEMORY or by defining nocache regions in devicetree. Use these Kconfigs to detect the devicetree-defined nocache region and to obtain its name.

erwango replied:
We likely inherit this heterogeneity of configurations from the various evolutions about DMA and cache handling.
Maybe some of them don't make sense anymore and maintaining all these configurations will be costly in the end. At one point we'll have to revisit this, and minimize the number of possible configurations.
I guess you now have a good view on this, don't hesitate to propose a way forward whenever you feel appropriate (you don't have to do it yourself though).

dgastonochoa replied:
It depends on which path Zephyr wants to follow. If the `CONFIG_NOCACHE_MEMORY` strategy is to be discarded, then I would delete all code related to it in the driver, and do either manual cache coherency, or add an extra **mandatory** Kconfig to the STM32 SPI driver to specify which nocache area it is expected to use (and assume the user has declared such area as nocache in devicetree). The problem with this is that not all STM32 SPI drivers need to care about this, so possibly such section should be mandatory for SPI STM32 when the current arch. has DCACHE. There are Kconfigs to detect this already, so it won't be very difficult I think.

On the other hand,  if  `CONFIG_NOCACHE_MEMORY` is the way to go, I would modify the stm32 SPI Kconfig to always select this feature. Then, go to the driver's source code and remove any logic related to handle the possibility of it not being enabled, and stop taking into account other nocache areas declared in devicetree.

Finally, I would remove from the SPI loopback tests the code related to the option discarded above.

The main point is deciding which way to go. It seems to me the devicetree option is the way to go, because devicetree is the place to describe how the HW is, and the user should have freedom to decide where to put this or that nocache buffer. Although this option seems more complex than just enabling a Kconfig.

erwango replied:
I think the direction is to remove `CONFIG_NOCACHE_MEMORY`, but there may have some blockers right now, this is tbc. Anyway this is not directly related to your PR.
Let me put this point as resolved so it doesn't this PR to progress.

dgastonochoa commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -582,51 +597,6 @@ ZTEST(spi_loopback, test_spi_loopback)
 		goto end;
 	}
 
-#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi)
```
To reduce complexity in this file, instead of having another set of tests for 16 bits frames, use overlay and conf files (as DMA does)

erwango commented on tests/drivers/spi/spi_loopback/overlay-stm32-spi-16bits-dma-no-nocache.conf:
```diff
@@ -0,0 +1,14 @@
+
+# enable DMA mode for SPI loopback test
+CONFIG_SPI_STM32_DMA=y
+# CONFIG_SPI_STM32_INTERRUPT is not set
+# CONFIG_SPI_ASYNC is not set
```
Please use :
```
CONFIG_SPI_STM32_INTERRUPT=n
CONFIG_SPI_ASYNC=n
```
This will be consistent with other .conf files and will take care of cases where symbols are defined in board config for instance.

dgastonochoa replied:
Done.

erwango commented on tests/drivers/spi/spi_loopback/overlay-stm32-spi-16bits.conf:
```diff
@@ -0,0 +1,6 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+CONFIG_SPI_LOOPBACK_16BITS_FRAMES=y
```
Do you expect some specific value for CONFIG_SPI_STM32_INTERRUPT ?
In this case, I think it should be specified.

FYI, there is a bit of inconsistency on STM32 SPI interrupt config setting.
It is enabled in lot of board configurations, but not all. We need to fix this at one point.

dgastonochoa replied:
> Do you expect some specific value for CONFIG_SPI_STM32_INTERRUPT ?

This configuration file intends to test non-DMA, non-interrupts SPI operations, so I specifically want it to be disabled.

> In this case, I think it should be specified.

It's now `CONFIG_SPI_STM32_INTERRUPT=n`, I imagine that's what you were referring to?

erwango commented on tests/drivers/spi/spi_loopback/overlay-stm32-spi-dma-no-nocache.conf:
```diff
@@ -0,0 +1,7 @@
+# enable DMA mode for SPI loopback test
+CONFIG_SPI_STM32_DMA=y
+# CONFIG_SPI_STM32_INTERRUPT is not set
+# CONFIG_SPI_ASYNC is not set
```
Ditto

erwango reviewed (changes requested):
Some comments otherwise looks good.
Also please split at least the code vs tests changes (mainly in order to avoid issues if test changes have impact on other platforms)

dgastonochoa committed:
drivers: spi: stm32h7 Fix SPI DMA 16 bits frames

Modify STM32H7 SPI driver so that it updates the rx/tx
pointers correctly (depending on the frame size) when DMA
is enabled. Also, make the dummy rx/tx buffer cache-coherent.

Signed-off-by: Daniel Gaston Ochoa <dgastonochoa@gmail.com>

dgastonochoa committed:
tests: drivers: spi: stm32h7: Run 16-bit frames tests for DMA

Run 16-bit frames tests also when DMA is enabled. In addition,
remove the extra code required for the mentioned tests in
favor of config. files, and add test cases for nocache regions
defined in devicetree.

Signed-off-by: Daniel Gaston Ochoa <dgastonochoa@gmail.com>

dgastonochoa commented:
> Also please split at least the code vs tests changes (mainly in order to avoid issues if test changes have impact on other platforms)

Done

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -28,10 +28,23 @@ LOG_MODULE_REGISTER(spi_ll_stm32);
 
 #include "spi_ll_stm32.h"
 
+/* Check defined(CONFIG_DCACHE) because some platforms disable it in the tests
+ * e.g. nucleo_f746zg
```
It could be interesting to find out why and fix it if it has no sense anymore.

dgastonochoa replied:
I'm afraid I don't have access to f746zg, and so I'm unable to test this.

dgastonochoa commented:
> I've run twister tests on `tests/drivers/spi` and `tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_dma_no_nocache.loopback, nucleo_h743zi ` is reported failed. Can you have a check ?

I've run them locally and they pass. This is exactly what I am doing:

```
rm -rf ~/zephyrproject/zephyr/build
west build -p auto -b nucleo_h753zi -T tests/drivers/spi/spi_loopback/drivers.spi.stm32_spi_dma_no_nocache.loopback
west flash
```

Could you provide the exact command you run, please?

erwango commented:
I'm running twister on our bench:
`twister --ninja --runtime-artifact-cleanup -e logging -e cmsis_dsp -T tests/drivers/spi --hardware-map <path>/../map.yaml --device-testing --outdir <path>`
Here is the log:
```
*** Booting Zephyr OS build zephyr-v3.4.0-3629-gf049ac191657 ***
Running TESTSUITE spi_loopback
===================================================================
START - test_spi_loopback
I: SPI test on buffers TX/RX 0x300200a0/0x30020080, frame size = 8, DMA enabled (without CONFIG_NOCACHE_MEMORY)
I: SPI test slow config
I: Start complete multiple
E: Code -14
Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:130: spi_complete_multiple: (ret is true)
SPI transceive failed
FAIL - test_spi_loopback in 0.029 seconds
===================================================================
TESTSUITE spi_loopback failed.
------ TESTSUITE SUMMARY START ------
SUITE FAIL -   0.00% [spi_loopback]: pass = 0, fail = 1, skip = 0, total = 1 duration = 0.029 seconds
- FAIL - [spi_loopback.test_spi_loopback] duration = 0.029 seconds
------ TESTSUITE SUMMARY END ------
===================================================================
RunID: 0e307fc99387fa5a6cab18de64a59ae0
PROJECT EXECUTION FAILED

```

dgastonochoa commented:
Not sure how to exec. that command in my machine; I don't have a `map` file to provide to `--hardware-map`. Is there any other way to indicate `twister` how to flash the result binaries? I do it with `west flash`.

erwango commented:
> Not sure how to exec. that command in my machine; I don't have a `map` file to provide to `--hardware-map`. Is there any other way to indicate `twister` how to flash the result binaries? I do it with `west flash`.

I don't think you should too much care about the map file, it's just a way to route the test to the correct board.
You could use `--device-testing --device-serial /dev/ttyACM0 -p board-name`

dgastonochoa commented:
Thanks.

This is the result I obtain:

```
$ west twister --ninja --runtime-artifact-cleanup -e logging -e cmsis_dsp -T tests/drivers/spi --device-serial /dev/ttyACM0 --device-testing -p nucleo_h753zi --outdir outdir


INFO    - Using Ninja..
INFO    - Zephyr version: zephyr-v3.4.0-3629-gf049ac191657
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report outdir/testplan.json

Device testing on:

| Platform      | ID   | Serial device   |
|---------------|------|-----------------|
| nucleo_h753zi |      | /dev/ttyACM0    |

INFO    - JOBS: 12
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:   18/  18  100%  skipped:   17, failed:    0, error:    0
INFO    - 18 test scenarios (18 test instances) selected, 17 configurations skipped (17 by static filter, 0 at runtime).
INFO    - 1 of 18 test configurations passed (100.00%), 0 failed, 0 errored, 17 skipped with 0 warnings in 14.20 seconds
INFO    - In total 1 test cases were executed, 34 skipped on 1 out of total 615 platforms (0.16%)
INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.

Hardware distribution summary:

| Board         | ID   |   Counter |
|---------------|------|-----------|
| nucleo_h753zi |      |         1 |
INFO    - Saving reports...
INFO    - Writing JSON report /home/danielgo/zephyrproject/zephyr/outdir/twister.json
INFO    - Writing xunit report /home/danielgo/zephyrproject/zephyr/outdir/twister.xml...
INFO    - Writing xunit report /home/danielgo/zephyrproject/zephyr/outdir/twister_report.xml...
INFO    - Run completed
```

It seems to pass, am I wrong?

erwango commented:
> It seems to pass, am I wrong?

Right. Can it come from a configuration difference between 2 the targets ?

dgastonochoa commented:
> Right. Can it come from a configuration difference between 2 the targets ?

Yes, it's that. I miss adding this:

```
&sram2 {
	zephyr,memory-attr = "RAM_NOCACHE";
};
```

To the H743 overlay file, thanks.

It should pass your test bench now I believe.

erwango commented:
> It should pass your test bench now I believe.

I confirm

dgastonochoa committed:
drivers: spi: stm32h7 Check nocache memory correctly

The way of setting a nocache region in devicetree has changed.
Adapt the H7 SPI driver to this new circumstance.

Signed-off-by: Daniel Gaston Ochoa <dgastonochoa@gmail.com>

dgastonochoa commented:
The memory management subsystem has been merged and so the way of detecting when a pointer is in a nocache area. Thus, I have needed to add a new commit to adapt the SPI driver to this situation @erwango

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -771,11 +746,13 @@ static bool buf_in_nocache(uintptr_t buf, size_t len_bytes)
 
-	for (size_t i = 0; i < ARRAY_SIZE(nocache_mem_regions); i++) {
-		const struct mem_region *mem_reg = &nocache_mem_regions[i];
+	const struct mem_attr_region_t *regions;
+	size_t num_regions =  mem_attr_get_regions(&regions);
+
+	for (size_t i = 0; i < num_regions; i++) {
```
Do we really need to iterate across regions ?
From my understanding, `mem_attr_check_buf()` will iterate across the array of regions.

Then, I didn't had coffee so maybe I'm not woken up enough, but if iteration is required, I wonder how it happens in your code

dgastonochoa replied:
Oops, yes, this loop is completely unnecessary. Fixed now, thanks.
